### ISSUE TITLE
ID scheme allows non-unique IDs

### ISSUE THREAD

**Issue by david4096** [ga4gh/server#1012]
Identifiers are required to be unique, but naming two files the same will cause them to have the same ID.

```
import ga4gh.datamodel.variants as variants
import ga4gh.datamodel.reads as reads
import ga4gh.datamodel.datasets as datasets

ds = datasets.AbstractDataset("bruce")
rgs = reads.AbstractReadGroupSet(ds, "wayne")

rgs.getId()
# 'YnJ1Y2U6d2F5bmU'

vs = variants.AbstractVariantSet(ds, "wayne")

vs.getId()
#'YnJ1Y2U6d2F5bmU'
```

Perhaps we could add a character per type to the compound ID?

**Comment by jeromekelleher** [comment]
Hah, well spotted @david4096! Yes, adding a char for each type is a good solution I think.

**Comment by dcolligan** [comment]
It is a bit ugly that 2 different objects of different type can share the same id, but is it really a problem, strictly speaking?  We don't have any getObjectById function (or something similar) which searches the global id space.  We only have type-specific id spaces, I think.

**Comment by jeromekelleher** [comment]
Hmm, this is true @dcolligan, we can always identify the correct object by the context. Does the spec have anything to say about this?

**Comment by david4096** [comment]
Although I agree that we can function with degenerate IDs it is out of scope of the schemas.
```
* The standard way to programmatically reference an object is via an
  **id** field. The id is server-assigned, and must provide “id-like”
  semantics, including:

  * ids are unique within the scope of the server instance
...
The purpose of IDs is to allow
unique identification of all objects within a single server, such that
no two objects in a given server have the same ID and no object has
more than one ID.
```
https://raw.githubusercontent.com/ga4gh/schemas/master/doc/source/api/apidesign_intro.rst

**Comment by jeromekelleher** [comment]
OK, that sounds conclusive. I think we should just use @david4096's idea and shove in a single char to make the IDs unique so we're following the letter of the spec.

**Comment by dcolligan** [comment]
+1

**Comment by diekhans** [comment]
From a user perspective, it's very confusing to not have ids
globally unique.  For debugging and if one does anything like a
cache.

Please figure out a short-term fix.  Once we get protocol
buffers integrated, it would be good to revisit the id scheme.

**Comment by macieksmuga** [comment]
The following is a sketch of a solution architecture that I'd like to implement at some point in the future, not a reflection on the code that's generating these IDs now.

Ideally, the ID conundrum would be solved by every class for an ID-backed entity having two required methods, the implementation of which would presumably call on some standard obfuscation utility:

```
getId(self) -> string

@classmethod
getById(string: id) ->  object representing the entity
```

The issue with this simple approach is that  server architecture assigns IDs to two distinct types of entities (from the server's viewpoint):

- data containers, such as datasets, readGroupSets, featureSets, variantSets, etc... - these correspond to directories and files that the server has to access in order to serve out their contents.
- data items and containers, such as variants, features, readGroups (which don't correspond to a container, but are entities named within a readGroupSet) - that correspond to data accessible within some container. 

Thus, given a request with a data item ID, the server needs to figure out the appropriate data container to access first. 

This could be done through a third standard `@classmethod` on every data item entity class that, given an ID provides the appropriate parent container, or possibly by folding this functionality automatically into the `getById` method. Some generic mechanisms need to be built to allow for all this, but I don't think that building and maintaining an entire parallel hierarchy of classes within `datamodel.__init__` just to satisfy this unique ID requirement is the right approach going forward.

**Comment by macieksmuga** [comment]
I created a document capturing the above architecture sketch for future reference:
https://docs.google.com/a/soe.ucsc.edu/document/d/1xhzM3_jGL0OgpEElok-5WGsdlfUxQ6KjF78jaFOdIvw/edit?usp=sharing

**Comment by dcolligan** [comment]
Fixed in #1060

(although the issue will probably need to be solved more generally somewhere down the road)
